oak_functions/loader/src/server.rs

```diff
@@ -153,15 +153,26 @@ pub trait ExtensionFactory {
 /// `Uwabi` extension called by listening to a channel.
 pub enum BoxedExtension {
     Native(Box<dyn OakApiNativeExtension + Send + Sync>),
-    Uwabi(Box<dyn UwabiExtension + Send + Sync>),
+    Uwabi(BoxedUwabiExtension),
 }
 
+pub type BoxedUwabiExtension = Box<dyn UwabiExtension + Send + Sync>;
+
 pub type BoxedExtensionFactory = Box<dyn ExtensionFactory + Send + Sync>;
 
 /// Trait for implementing an extension which relies on UWABI.
 pub trait UwabiExtension {
     /// Get the channel handle to address this extension.
     fn get_channel_handle(&self) -> ChannelHandle;
+
+    /// Get the endpoint.
+    fn get_endpoint(&mut self) -> Option<&mut Endpoint>;
+
+    /// Set the endpoint if it has not been set before.
+    // Note: we cannot set the endpoint when we `create` the extension, as this would require
```

Comment:
It seems that the existing factory trait is already being stretched by returning two completely different types, and now it seems that it should also have different inputs. Have you considered just having a different factory trait, instead of trying to share the previous one? That may also simplify other things that currently cannot be enforced because of the ambiguity.

Reply:
Adding a different factory trait would require me to change the [`WasmHandler`](https://github.com/project-oak/oak/blob/f1e9783cb872e9df58cce5a6d74ec8e872eb3822/oak_functions/loader/src/server.rs#L662), which I am trying to avoid. 

I think, having this two extension---`Uwabi` and `Native` is the simplest solution for now---leaving most of the current functionality untouched.